src/transform/src/redundant_join.rs

```diff
@@ -341,28 +348,110 @@ impl RedundantJoin {
 pub struct ProvInfo {
     // The Id (local or global) of the source.
     id: Id,
-    // A list of (source column, current column) associations.
-    // There should be at most one occurrence of each number in the second position.
-    binding: Vec<(usize, usize)>,
+    // The projection of the current operator written in terms of the columns
+    // projected by the underlying Get operator.
+    dereferenced_projection: Vec<Option<MirScalarExpr>>,
     // If true, all distinct projected source rows are present in the rows of
     // the projection of the current collection. This constraint is lost as soon
     // as a transformation may drop records.
     exact: bool,
 }
 
 impl ProvInfo {
+    fn make_leaf(id: Id, arity: usize) -> Self {
+        Self {
+            id,
+            dereferenced_projection: (0..arity)
+                .map(|c| Some(MirScalarExpr::column(c)))
+                .collect::<Vec<_>>(),
+            exact: true,
+        }
+    }
+
+    /// Given an expression on top of the projection of the current operator
+    /// it returns an equivalent expression as if it was right on top of the
+    /// underlying Get operator.
+    fn dereference(&self, expr: &MirScalarExpr) -> Option<MirScalarExpr> {
```

Comment:
Maybe we can add `try_visit_mut` (see the corresponding method in src/expr/src/relation/mod.rs) and use `try_visit_mut` in this method and in `try_build_expression_using_other`. Whenever, this method would have returned None, throw an error. But then again, using `try_visit_mut` would require cloning `expr`.